<a href="https://colab.research.google.com/github/Yangyang5750858/yangyang5750858-DataScience-GenAI-Submissions/blob/main/Assignment_5/6_02_DNN_101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://drive.google.com/uc?export=view&id=1xqQczl0FG-qtNA2_WQYuWePW9oU8irqJ)

# 6.02 Dense Neural Network (with PyTorch)
This will expand on our logistic regression example and take us through building our first neural network. If you haven't already, be sure to check (and if neccessary) switch to GPU processing by clicking Runtime > Change runtime type and selecting GPU. We can test this has worked with the following code:

In [ ]:
import torch

# Check for GPU availability
print("Num GPUs Available: ", torch.cuda.device_count())

Num GPUs Available:  1


Hopefully your code shows you have 1 GPU available! Next let's get some data. We'll start with another in-built dataset:

In [ ]:
# upload an in-built Python (OK semi-in-built) dataset
from sklearn.datasets import load_diabetes

import pandas as pd
import numpy as np

# import the data
data = load_diabetes()
data

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990749, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06833155, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286131, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04688253,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452873, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00422151,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

We are working on a regression problem, with "structured" data which has already been cleaned and normalised. We can skip the usual cleaning/engineering steps. However, we do need to get the data into PyTorch:

In [ ]:
# Convert data to PyTorch tensors
X = torch.tensor(data.data, dtype=torch.float32)
y = torch.tensor(data.target, dtype=torch.float32).reshape(-1, 1) # Reshape y to be a column vector

Now our data is stored in tensors we can do train/test splitting as before (in fact we can use sklearn as before):

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

torch.Size([353, 10]) torch.Size([353, 1])
torch.Size([89, 10]) torch.Size([89, 1])


Now we can set up our batches for training. As we have a nice round 400 let's go with batches of 50 (8 batches in total). We'll also seperate the features and labels:

In [ ]:
from torch.utils.data import TensorDataset, DataLoader

# Create TensorDatasets and DataLoaders
train_dataset = TensorDataset(X_train, y_train)
train_loader = DataLoader(train_dataset, batch_size=50, shuffle=True)

test_dataset = TensorDataset(X_test, y_test)
test_loader = DataLoader(test_dataset, batch_size=50, shuffle=False)

Now its time to build our model. We'll keep it simple ... a model with an input layer of 10 features and then 2x _Dense_ (fully connected) layers each with 5 neurons and ReLU activation. Our output layer will be size=1 given this is a regression problem and we want a single value output per prediction.

This will be easier to understand if you have read through the logistic regression tutorial.

In [ ]:
import torch
import torch.nn as nn

# Define the model
class DiabetesModel(nn.Module):
    def __init__(self):
        super(DiabetesModel, self).__init__()
        # we'll set up the layers as a sequence using nn.Sequential
        self.layers = nn.Sequential(

            # first layer will be a linear layer that has 5x neurons
            # (5x sets of linear regression)
            # the layer takes the 10 features as input (i.e. 10, 5)
            nn.Linear(10, 5),

            nn.ReLU(), # ReLU activation

            # second linear layer again has 5 neurons
            # this time taking the input as the output of the last layer
            # (which had 5x neurons)
            nn.Linear(5, 5),

            nn.ReLU(), # ReLU again

            # last linear layer takes the output from the previous 5 neurons
            # this time its a single output with no activation
            # i.e. this is the predicitons (regression)
            nn.Linear(5, 1)
        )

    def forward(self, x):
        return self.layers(x) # pass the data through the layers

As before we need to create a model object, specify the loss (criterion) and an optimiser (which we cover next week):

In [ ]:
import torch.optim as optim

# Initialize the model, loss function, and optimizer
model = DiabetesModel()
criterion = nn.MSELoss() # MSE loss function
optimiser = optim.Adam(model.parameters(), lr=0.001)

Now we can train the model. Again, the logistic regression tutorial (6.01) may help you undertstand this:

In [ ]:
# Move model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Training loop (example - you'll likely want to add more epochs)
epochs = 100 # 100 epochs

for epoch in range(epochs):
  # use the train_loader to pass the inputs (x) and targets (y)
  for inputs, targets in train_loader:
    # pass to the GPU (hopefully)
    inputs, targets = inputs.to(device), targets.to(device)

    # pass model to GPU as well
    model.to(device)

    model.train() # put the model object in train mode
    optimiser.zero_grad() # reset the gradiants
    outputs = model(inputs) # create outputs
    loss = criterion(outputs, targets) # compare with Y to get loss
    loss.backward() # backpropogate the loss (next week)
    optimiser.step() # # update the parameters based on this round of training

  # every 10 steps we will print out the current loss
    if (epoch+1) % 10 == 0: # modular arithmetic
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {round(loss.item(), 4)}')

Epoch [10/100], Loss: 24366.6289
Epoch [10/100], Loss: 34809.3945
Epoch [10/100], Loss: 23924.1211
Epoch [10/100], Loss: 30655.3965
Epoch [10/100], Loss: 24983.9629
Epoch [10/100], Loss: 35090.168
Epoch [10/100], Loss: 35043.6562
Epoch [10/100], Loss: 4565.8086
Epoch [20/100], Loss: 34702.625
Epoch [20/100], Loss: 31930.5273
Epoch [20/100], Loss: 28059.291
Epoch [20/100], Loss: 30386.7246
Epoch [20/100], Loss: 27855.3945
Epoch [20/100], Loss: 28101.0195
Epoch [20/100], Loss: 25984.6602
Epoch [20/100], Loss: 29510.8105
Epoch [30/100], Loss: 34797.1992
Epoch [30/100], Loss: 29668.8203
Epoch [30/100], Loss: 31846.2461
Epoch [30/100], Loss: 26075.291
Epoch [30/100], Loss: 27496.4199
Epoch [30/100], Loss: 26569.334
Epoch [30/100], Loss: 30520.9668
Epoch [30/100], Loss: 23558.123
Epoch [40/100], Loss: 29638.7422
Epoch [40/100], Loss: 36933.4805
Epoch [40/100], Loss: 31415.127
Epoch [40/100], Loss: 29869.7188
Epoch [40/100], Loss: 30518.4336
Epoch [40/100], Loss: 24049.3086
Epoch [40/100], Lo

We can see loss is significantly lower at the end than it was at the start. However, it is also bouncing around a little still which suggests the model needs more training (100 epochs is not a lot in deep learning terms). However, let's evaluate as before:

In [ ]:
# Evaluation (example)
model.eval() # testing mode
mse_values = [] # collect the MSE scores

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs) # predict the test data

        # Calculate Mean Squared Error
        mse = criterion(outputs, targets) # calcualte mse for the batch
        mse_values.append(mse.item()) # add to the list of MSE values

# Calculate and print the average MSE
avg_mse = np.mean(mse_values)
print(f"Average MSE on test set: {avg_mse}")

Average MSE on test set: 25403.248046875


MSE looks expected given training (no obvious sign of overfitting). However, we probably can get better results with tuning and more epochs.

Let's run the loop again a little differently to collect the predicted values (y_hat) and actuals (y) and add them to a dataset for comparions:

In [ ]:
# Evaluation
model.eval()
predictions = []
actuals = []

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs)
        predictions.extend(outputs.cpu().numpy())
        actuals.extend(targets.cpu().numpy())

# Create DataFrame
results_df = pd.DataFrame({'Predicted': np.array(predictions).flatten(), 'Actual': np.array(actuals).flatten()})
results_df

Predicted  Actual
0    2.718976   219.0
1    2.718976    70.0
2    2.718976   202.0
3    2.718976   230.0
4    2.718976   111.0
..        ...     ...
84   2.718976   153.0
85   2.718976    98.0
86   2.718976    37.0
87   2.718976    63.0
88   2.718976   184.0

[89 rows x 2 columns]

Side-by-side, they don't look great. Can you improve them?

<br><br>

## EXERCISE #1
Try increasing the number of epochs to 1,000 (when the model is fairly well trained then the results printed for each 10x epochs will be fairly stable and not change much). Does this give better results?

<br><br>

## EXERCISE #2 (optional)
Try experimenting with the architecture (number of neurons and/or number of layers). Can we reach an optimal architecture?

Exercise 1: what we have to do now is to change the epochs, so lets copy the code of the previous sections and see the outcomes.

In [ ]:
epochs = 1000 # We need to change this row

for epoch in range(epochs):
  for inputs, targets in train_loader:
    inputs, targets = inputs.to(device), targets.to(device)

    model.to(device)

    model.train()
    optimiser.zero_grad()
    outputs = model(inputs)
    loss = criterion(outputs, targets)
    loss.backward()
    optimiser.step()

    if (epoch+1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {round(loss.item(), 4)}')

Epoch [10/1000], Loss: 31479.7051
Epoch [10/1000], Loss: 23057.9766
Epoch [10/1000], Loss: 27466.8145
Epoch [10/1000], Loss: 26555.5742
Epoch [10/1000], Loss: 34904.0469
Epoch [10/1000], Loss: 28379.0176
Epoch [10/1000], Loss: 29363.6484
Epoch [10/1000], Loss: 31721.5703
Epoch [20/1000], Loss: 33947.8945
Epoch [20/1000], Loss: 25497.6035
Epoch [20/1000], Loss: 30965.8398
Epoch [20/1000], Loss: 24973.8145
Epoch [20/1000], Loss: 27979.9844
Epoch [20/1000], Loss: 27635.7793
Epoch [20/1000], Loss: 26899.0293
Epoch [20/1000], Loss: 71503.25
Epoch [30/1000], Loss: 25953.0117
Epoch [30/1000], Loss: 26720.7793
Epoch [30/1000], Loss: 32737.7246
Epoch [30/1000], Loss: 33376.9727
Epoch [30/1000], Loss: 25494.5938
Epoch [30/1000], Loss: 25483.2461
Epoch [30/1000], Loss: 29601.709
Epoch [30/1000], Loss: 30673.5117
Epoch [40/1000], Loss: 29877.2539
Epoch [40/1000], Loss: 32469.7148
Epoch [40/1000], Loss: 30722.8242
Epoch [40/1000], Loss: 23731.6211
Epoch [40/1000], Loss: 31414.3691
Epoch [40/1000], 

In [ ]:
# Lets see now the average MSE
model.eval()
mse_values = []

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs)

        mse = criterion(outputs, targets)
        mse_values.append(mse.item())

# Calculate and print the average MSE
avg_mse = np.mean(mse_values)
print(f"Average MSE on test set: {avg_mse}")

Average MSE on test set: 6372.0234375


In [ ]:
# Now lets copare the predicted data and the actual
model.eval()
predictions = []
actuals = []

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs)
        predictions.extend(outputs.cpu().numpy())
        actuals.extend(targets.cpu().numpy())

results_df = pd.DataFrame({'Predicted': np.array(predictions).flatten(), 'Actual': np.array(actuals).flatten()})
results_df

Predicted  Actual
0   110.337082   219.0
1   110.337082    70.0
2   110.337082   202.0
3   110.337082   230.0
4   110.337082   111.0
..         ...     ...
84  110.337082   153.0
85  110.337082    98.0
86  110.337082    37.0
87  110.337082    63.0
88  110.337082   184.0

[89 rows x 2 columns]

After chanching the epoch into 1000 and training the model, we can see that it improved in comparison to epoch = 100.

The average MSE on the test set was approximately 6372.02, better than before but  still indicates that the model has a high prediction error.

When examining the “Predicted vs Actual” table, it can be seen that most of the predicted values are very close to 110.33, regardless of the true target value.

Could be that network has not properly learned the mapping between input features and the output variable. Possible reasons include: the model being too simple, too few training epochs or too low learning rate.